server/storage/bootstrap.go

```diff
@@ -0,0 +1,335 @@
+// Copyright 2021 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/dustin/go-humanize"
+	"go.uber.org/zap"
+
+	"go.etcd.io/etcd/api/v3/etcdserverpb"
+	"go.etcd.io/etcd/client/pkg/v3/fileutil"
+	"go.etcd.io/etcd/client/pkg/v3/types"
+	"go.etcd.io/etcd/pkg/v3/pbutil"
+	"go.etcd.io/etcd/raft/v3"
+	"go.etcd.io/etcd/raft/v3/raftpb"
+	"go.etcd.io/etcd/server/v3/config"
+	"go.etcd.io/etcd/server/v3/etcdserver/api/snap"
+	"go.etcd.io/etcd/server/v3/etcdserver/api/v2store"
+	"go.etcd.io/etcd/server/v3/etcdserver/cindex"
+	"go.etcd.io/etcd/server/v3/storage/backend"
+	"go.etcd.io/etcd/server/v3/storage/schema"
+	"go.etcd.io/etcd/server/v3/storage/wal"
+	"go.etcd.io/etcd/server/v3/storage/wal/walpb"
+)
+
+type BootstrappedStorage struct {
+	Backend *bootstrappedBackend
+	Wal     *BootstrappedWAL
+	St      v2store.Store
+}
+
+type bootstrappedBackend struct {
+	BeHooks  *BackendHooks
+	Be       backend.Backend
+	Ci       cindex.ConsistentIndexer
+	BeExist  bool
+	Snapshot *raftpb.Snapshot
+}
+
+type BootstrappedWAL struct {
+	lg *zap.Logger
+
+	HaveWAL  bool
+	Wal      *wal.WAL
+	State    *raftpb.HardState
+	Entries  []raftpb.Entry
+	Snapshot *raftpb.Snapshot
+	Meta     *Metadata
+}
+
+type Metadata struct {
+	NodeID, ClusterID types.ID
+}
+
+func BootstrapStorage(cfg config.ServerConfig, st v2store.Store, be *bootstrappedBackend, wal *BootstrappedWAL, meta *Metadata) (b *BootstrappedStorage, err error) {
+	if wal == nil {
+		wal = bootstrapNewWAL(cfg, meta)
+	}
+
+	return &BootstrappedStorage{
+		Backend: be,
+		St:      st,
+		Wal:     wal,
+	}, nil
+}
+
+func BootstrapBackend(cfg config.ServerConfig, st v2store.Store, ss *snap.Snapshotter) (backend *bootstrappedBackend, err error) {
+	beExist := fileutil.Exist(cfg.BackendPath())
+	ci := cindex.NewConsistentIndex(nil)
+	beHooks := NewBackendHooks(cfg.Logger, ci)
+	be := OpenBackend(cfg, beHooks)
+	defer func() {
+		if err != nil {
+			be.Close()
+		}
+	}()
+	ci.SetBackend(be)
+	schema.CreateMetaBucket(be.BatchTx())
+	if cfg.ExperimentalBootstrapDefragThresholdMegabytes != 0 {
+		err = maybeDefragBackend(cfg, be)
+		if err != nil {
+			return nil, err
+		}
+	}
+	cfg.Logger.Debug("restore consistentIndex", zap.Uint64("index", ci.ConsistentIndex()))
+	var (
+		walSnaps []walpb.Snapshot
+		snapshot *raftpb.Snapshot
+	)
+
+	if beExist {
```

Comment:
Why did we have to merge this methods ? 

Method that does not fit on 4K display for me is a warning signal ;)

Reply:
Goal of this change was to consolidate code operating on backend to one place. Merging of function was just to simplify diff and review. As this code overrides backend be variable it's easy to make mistake due to different bracketing/function stack. 

I would be happy to split larger function into smaller but it's pretty delicate code and this was not my goal in this refactor.